### PR TITLE
updated files with my JUnit assignment

### PR DESCRIPTION
**Pull request by kaimmej**

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,32 +1,107 @@
...
+        String[] words = s.split("\\s+");
+        return words.length;
     }
 
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
+
+        if(s == null){
+            throw new NullPointerException("string cannot be null");
```
Comment (mihaisiia) [inline src/main/java/StringManipulation.java:43]: Null ptr exception not mentioned in the interface for the assigment.
Throw indexoutofbound? null str has no words

File: src/test/java/StringManipulationTest.java [hunk 2]
```diff
@@ -20,66 +29,87 @@ public void tearDown() {
     }
 
     @Test
-    public void testCount1() {
+    public void testCount1_NormalString() {
```
Comment (mihaisiia) [inline src/test/java/StringManipulationTest.java:32]: Renamed test case provided by Sara

File: src/test/java/StringManipulationTest.java [hunk 3]
```diff
@@ -20,66 +29,87 @@ public void tearDown() {
...
-        fail("Not yet implemented");
+    public void testCount4_TabSeperatedWords() {
+        manipulatedstring.setString("This   is    my    string");
+        int length = manipulatedstring.count();
+        assertEquals(4, length);
     }
 
     @Test
-    public void testRemoveNthCharacter1() {
+    public void testRemoveNthCharacter1_DontMaintainSpacing() {
```
Comment (mihaisiia) [inline src/test/java/StringManipulationTest.java:60]: Renamed test case provided by sara

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,32 +1,107 @@
...
-    public String restoreString(int[] indices) {
-        return null;
+    public String restoreString(int[] indices){
+        if (s.length() != indices.length) {
+            throw new IllegalArgumentException("The length of the string and the indices array must be the same.");
+        }
+
+        char[] shuffled = new char[s.length()];
+
+        for (int i = 0; i < s.length(); i++) {
```
Comment (mihaisiia) [inline src/main/java/StringManipulation.java:97]: move exception checking up? save on shuffling in event that exception is thrown later? idk may not be computationally significant i'm just writing words

File: src/test/java/StringManipulationTest.java [hunk 5]
```diff
@@ -20,66 +29,87 @@ public void tearDown() {
     }
 
     @Test
```
Comment (mihaisiia) [inline src/test/java/StringManipulationTest.java:31]: what if string contains non alphanumeric chars? no test case but maybe that's not even in the scope of this assignment who knows?

**Review by mihaisiia — APPROVED** [review]

File: src/main/java/StringManipulation.java (hunk 1 above)
Comment (kaimmej) [reply src/main/java/StringManipulation.java:43]: Yeah you are probably right, I just wanted to add exception checking and figured that a string might be null so why not check whether the string was nulll.. indexOutOfBounds is a good suggestion

File: src/test/java/StringManipulationTest.java (hunk 5 above)
Comment (kaimmej) [reply src/test/java/StringManipulationTest.java:31]: That edge case was out of the bounds of the assignment. 
Future improvement for other classes for sure

File: src/main/java/StringManipulation.java [hunk 6]
```diff
@@ -1,32 +1,107 @@
...
+    public String s;
+
     @Override
     public String getString() {
-        return null;
+        return s;
     }
 
     @Override
     public void setString(String string) {
```
Comment (mihaisiia) [inline src/main/java/StringManipulation.java:26]: In the event that user sets string = "", consider overriding that input with null? May be outside of how Sara expects us to implement, though perhaps worth considering for methods below. Save on calling isEmpty(). For sure up for interpretation, no single answer.

File: src/test/java/StringManipulationTest.java (hunk 2 above)
Comment (mihaisiia) [inline src/test/java/StringManipulationTest.java:32]: count() test cases provided for alphabet/empty/null strings, further test cases by Sara suggest string provided may include numbers interspersed with letters, consider adding a test case for this scenario.

Edit: I realized that this comment is just rehashing what I wrote in the comment above this one, but I will leave it up anyways.

File: src/test/java/StringManipulationTest.java [hunk 7]
```diff
@@ -88,28 +118,53 @@ public void testGeSubStrings1() {
...
-    public void testGeSubStrings6() {
-        fail("Not yet implemented");
+    public void testGetSubStrings6_EmptyString() {
+        manipulatedstring.setString("");
+
+        assertThrows(IndexOutOfBoundsException.class,
+                ()-> {
+                    String [] sStings = manipulatedstring.getSubStrings(1, 5);
+                });
     }
```
Comment (mihaisiia) [inline src/test/java/StringManipulationTest.java:165]: All RestoreString() test cases use the same string, consider adding tests that deal with numbers/symbols/whitespace.

File: src/test/java/StringManipulationTest.java [hunk 8]
```diff
@@ -20,66 +29,87 @@ public void tearDown() {
...
 
     @Test
-    public void testCount4() {
-        fail("Not yet implemented");
+    public void testCount4_TabSeperatedWords() {
+        manipulatedstring.setString("This   is    my    string");
+        int length = manipulatedstring.count();
+        assertEquals(4, length);
     }
```
Comment (mihaisiia) [inline src/test/java/StringManipulationTest.java:58]: How should removeNthCharacter(1, true/false) behave? Should the method return "" or null?

File: src/test/java/StringManipulationTest.java [hunk 9]
```diff
@@ -119,31 +174,42 @@ public void testRestoreString1()
...
-        fail("Not yet implemented");
+        manipulatedstring.setString("art");
+        int [] array;
+        array=new int[]{1,0,-10};
 
+        assertThrows(IndexOutOfBoundsException.class,
+                ()-> {
+                    String restoreString = manipulatedstring.restoreString(array);
+                });
     }
```
Comment (mihaisiia) [inline src/test/java/StringManipulationTest.java:213]: With the exception of testCount(), all of the test cases check if exceptions are thrown. It may be worth adding unique test cases that test your program logic to ensure that your methods operate as you intend them to.

File: src/main/java/StringManipulation.java [hunk 10]
```diff
@@ -1,32 +1,107 @@
...
+
+import java.lang.IllegalArgumentException;
+import java.lang.IndexOutOfBoundsException;
+import java.lang.NullPointerException;
+import java.util.*;
+
+
 public class StringManipulation implements StringManipulationInterface {
 
+    public String s;
```
Comment (mihaisiia) [inline src/main/java/StringManipulation.java:18]: Changing String from public to protected/private here would increase code encapsulation.

File: src/main/java/StringManipulation.java [hunk 11]
```diff
@@ -1,32 +1,107 @@
...
+            throw new NullPointerException("string cannot be null");
+        }
+        if (n <= 0) {
+            throw new IllegalArgumentException("n must be greater than zero.");
+        }
+        if (n > s.length()){
+            throw new IndexOutOfBoundsException("n is greater than the string length.");
+        }
+
+        StringBuilder sb = new StringBuilder();
```
Comment (mihaisiia) [inline src/main/java/StringManipulation.java:52]: Here you can instantiate StringBuilder() by passing an existing string to it. After, you can call StringBuilder.remove(int index) to remove characters at a given position. If maintainSpacing == true, StringBuilder.setCharAt(int index, char ch) can be called to replace chars with ' '.

File: src/main/java/StringManipulation.java
Comment (mihaisiia) [inline src/main/java/StringManipulation.java:1]: Hi Jon, I've added more comments in addition to the those added 6/16. Overall, your code is clean and easy to follow, with most of my comments reading more as differences of opinion than suggestions to change anything. I especially like how you leverage the java.util library to accomplish the objectives of the assignment. After the C++ we learned in past classes, I forgot how important builtin libraries are. While some may critique this as unnecessary code coupling, I truly see it as an excellent example of code reuse, saving dev time. The one major recommendation I have is to add more variety to the test cases. Otherwise, great work!

**Review by mihaisiia — APPROVED** [review]
